### ISSUE TITLE
Enforce fields by default when creating exceptions

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#3657]
Today, if I define:

    defmodule FooError do
      defexception [:message]
    end

And I call it as:

    # Notice the typo
    raise FooError, mesage: "foo bar"

It will raise the exception as usual and fail when showing the error message. We must change the default behaviour when raising exception to validate all fields. Users can customize by implementing `def exception/1` in their code.

**Comment by gjaldon** [comment]
@josevalim I'd like to take this on, if available. :)

**Comment by josevalim** [comment]
This is under discussion and it will land only after 1.2 so I would wait before work starts. :)

**Comment by gjaldon** [comment]
Ah, I see. I will wait until then. :) I think it makes sense to validate the options passed to `raise/2`. This reminds me of a typo I made for `:reloadable_paths` config in Phoenix. It took me a while to figure out that I misspelled it.

For Mix, I'm thinking we support a module_attribute like `@accepted_fields` (or some better name) where `config/2` and `config/3` will first check that attribute if the all the keys passed as opts are accepted.  Something like:

```elixir
@accepted_fields lager: [:log_level, :threshold], ecto: [Repo: [:log_level, :pool_size]]

config :lager,
   log_level: :warn

config :lager,
  log_level: :warn
  mode: :truncate    #=======> Raises error
```

WDYT?

**Comment by josevalim** [comment]
@gjaldon that should be discussed in another issue. :)

**Comment by josevalim** [comment]
@lexmag and @whatyouhide: I would appreciate your feedback on this one. If we implement this for now, we would need to start as a warning.

**Comment by whatyouhide** [comment]
@josevalim yep, I think it's a good idea to do this now and warn, can't see any downsides :). Would we move to `struct!/2` when we can?

**Comment by josevalim** [comment]
I am working on this. I have decided to emit a warning for any field that does not have a default value.

We cannot use `struct!` because it is slightly different. `struct!` raises for unknown options, here we want to warn/raise for missing options.

**Comment by josevalim** [comment]
We can also use `struct!` if we want. Maybe we should start emit warning for those cases too?

**Comment by whatyouhide** [comment]
@josevalim I think the point is to not use fields that are not supported by the exception struct, correct? (that's why I thought of `struct!/2` over the `struct/2` we use now)
